staging/kos/pkg/controllers/connectionagent/connectionagent.go

```diff
@@ -0,0 +1,1467 @@
+
+	// namespace is the namespace of the Virtual Network
+	// associated with this vnState.
+	namespace string
+
+	// localAtts and remoteAtts store the names of the local and remote
+	// NetworkAttachments in the Virtual Network the vnState represents,
+	// respectively. localAtts is used to detect when the last local attachment
+	// in the virtual network is deleted, so that remoteAttsInformer can be
+	// stopped. remoteAtts is used to enqueue references to the remote
+	// attachments in the Virtual Network when such Virtual Network becomes
+	// irrelevant (deletion of last local attachment), so that the interfaces of
+	// the remote attachments can be deleted.
+	localAtts  map[string]struct{}
+	remoteAtts map[string]struct{}
+}
+
+// ConnectionAgent represents a K8S controller which runs on every node of the
+// cluster and eagerly maintains up-to-date the mapping between virtual IPs and
+// physical IPs for every relevant NetworkAttachment. A NetworkAttachment is
+// relevant to a connection agent if: (1) it runs on the same node as the
+// connection agent, or (2) it's part of a Virtual Network where at least one
+// NetworkAttachment for which (1) is true exists. To achieve its goal, a
+// connection agent receives notifications about relevant NetworkAttachments
+// from the K8s API server through Informers, and when necessary
+// creates/updates/deletes Network Interfaces through a low-level network
+// interface fabric. When a new Virtual Network becomes relevant for the
+// connection agent because of the creation of the first attachment of that
+// Virtual Network on the same node as the connection agent, a new informer on
+// remote NetworkAttachments in that Virtual Network is created. Upon being
+// notified of the creation of a local NetworkAttachment, the connection agent
+// also updates the status of such attachment with its host IP and the name of
+// the interface which was created.
+type ConnectionAgent struct {
+	localNodeName string
+	hostIP        gonet.IP
+	kcs           *kosclientset.Clientset
+	netv1a1Ifc    netvifc1a1.NetworkV1alpha1Interface
+	eventRecorder k8seventrecord.EventRecorder
+	queue         k8sworkqueue.RateLimitingInterface
+	workers       int
+	netFabric     netfabric.Interface
+	stopCh        <-chan struct{}
+
+	// Informer and lister on NetworkAttachments on the same node as the
+	// connection agent
+	localAttsInformer k8scache.SharedIndexInformer
+	localAttsLister   koslisterv1a1.NetworkAttachmentLister
+
+	// Map from vni to vnState associated with that vni. Accessed only while
+	// holding vniToVnStateMutex
+	vniToVnStateMutex sync.RWMutex
+	vniToVnState      map[uint32]*vnState
+
+	// nsnToVNStateVNI maps local attachments namespaced names to the VNI of the
+	// vnState they're stored in. Accessed only while holding nsnToVNStateVNIMutex.
+	nsnToVNStateVNIMutex sync.RWMutex
+	nsnToVNStateVNI      map[k8stypes.NamespacedName]uint32
+
+	nsnToLocalStateMutex      sync.RWMutex
```

Comment:
I wonder if we could remove this mutex by putting in LocalAttachmentMainState some mutex-protected way to receive test results.